source/MaterialXGenShader/TypeDesc.h

```diff
@@ -153,49 +158,31 @@ class MX_GENSHADER_API TypeDesc
         }
     };
 
-    /// Return a type description by name.
-    /// If no type is found Type::NONE is returned.
-    static TypeDesc get(const string& name);
-
-    /// Remove a type description by name, if it exists.
-    static void remove(const string& name);
-
     static const string NONE_TYPE_NAME;
 
-    /// Create a Value from a string for a given typeDesc
+    /// Create a Value from a string for this TypeDesc
     ValuePtr createValueFromStrings(const string& value) const;
 
   private:
-    /// Simple constexpr hash function, good enough for the small set of short strings that
+    /// Simple hash function, good enough for the small set of short strings that
     /// are used for our data type names.
-    constexpr uint32_t constexpr_hash(std::string_view str, uint32_t n = 0, uint32_t h = 2166136261)
+    uint32_t string_hash(const string& str, uint32_t n = 0, uint32_t h = 2166136261)
     {
-        return n == uint32_t(str.size()) ? h : constexpr_hash(str, n + 1, (h * 16777619) ^ (str[n]));
+        return n == uint32_t(str.size()) ? h : string_hash(str, n + 1, (h * 16777619) ^ (str[n]));
     }
 
     uint32_t _id;
     uint8_t _basetype;
     uint8_t _semantic;
     uint16_t _size;
-    uint16_t _structIndex;
-};
-
-/// @class TypeDescRegistry
-/// Helper class for type registration.
-class MX_GENSHADER_API TypeDescRegistry
-{
-  public:
-    TypeDescRegistry(TypeDesc type, const string& name);
+    const string* _name;
```

Comment:
As we're only storing the raw pointer here, aren't we relying on this TypeDesc object only ever being used inside the lifetime of its respective GenContext object?  I'm not 100% sure on the life cycle of GenContext - but there are places where TypeDesc gets read from the ShaderPort in the MaterialXRenderXXX modules - is it possible that the GenContext object has been destroyed by that point?  I also wonder if there are life-cycle issues if TypeDesc objects are passed across the python binding boundary - and used who knows when.

Wondering if instead it's safer to store a shared_ptr<const string> here?  Which would let us do away with the string storage in GenContext too.... thoughts?

Reply:
That's a great point Lee. I used a raw pointer to minimize the class size still, and my plan was to change `_structMembers` to a raw pointer too for the same reason. As you know, the original design of TypeDesc was a minimal size POD class that could be passed by value and stored by value everywhere in the API. So using smart pointers here is not ideal, as they both take more space and will need refcount increment/decrement for every pass and use of a TypeDesc.

But your are definitely right about the possible lifetime issues when the data is now stored in the GenContext.

I think we have two options:

1: We continue in this direction and let TypeDesc be a "large" class. It can hold all data needed internally and control lifetime of this data, adding smart pointers and all that. But then I think the current API design of passing and storing these by value everywhere would need to change as well. At least passing them by reference instead. And maybe also storing themselves as smart pointers everywhere, on the ShaderPorts etc.

2: We revert back to the original design where TypeDesc is a minimal class, and we keep all API unchanged in terms of passing and storing them. Large data like name strings and struct members needs separate storage then, but since the lifetime of a GenContex is not guaranteed the GenContex may not be the best place to store this. So I wonder if it would be better to revert back to a singleton class managing all TypeDesc storage after all, and instead take a different approach for making the data access thread safe.

I think I'm in favour of option (2) as that would be a much smaller update and no API would need to change. Instead of using thread local storage (GenContext) we could add some synchronization logic for accessing TypeDesc data in the singleton storage. The data can then be accessed from everywhere and we don't have to make the API changes to pass the GenContext around either.

Note that access to the standard data types has always been thread-safe, as their TypeDesc's are statically created. So it's only the dynamic creation of struct types that adds the need for some extra thread safety. Thread synchronization using std::mutex/std::lock_guard is already used elsewhere in MaterialX so perhaps that's something we could use here as well.

What do you think? If you agree it's not a bad idea I could try this out in another branch, to see how it works out in practice. If it works out well we could merge the best parts from both branches.

Reply:
Thanks for those points Lee. There are certainly some advantages of letting the GenContext be the storage/maintainer of the type system. But there are also cons, and in particular two that I'm seeing. The first is that a lot of API changes are needed to pass around the GenContext everywhere in order to be able to query the types. And the second is the lifetime handling of the larger data. If TypeDesc's are used and needed after the lifetime of the GenContext it just doesn't feel right to let the GenContext be the owner of this data. Adding extra members to all ShaderPorts, in order to transfer this ownership in special situations, doesn't feel like a solid design to be honest. To my mind it's just so much cleaner to have a global type system, similar to what we have now, but make it thread safe to access everywhere. I think we can use a better separation of built-in types and custom/struct types in order to only need thread syncronization when custom types are queried, which is probably less than 1% of the cases.

I have a changelist in the works to better exaplain what I mean. Maybe we can take that discussion offline? I can ping you on Slack when I have something ready :)

Btw, your suggestion about trying to reduce our binding interfaces is something I agree on 100%.